Rework windows font collector

moi15moi opened the pull request:
The main goal of this PR is to replace the "hack" in the Windows FontCollector which involves using the Windows registry to obtain the font filename. Now, we use DirectWrite which can interop with GDI, so we are 100% sure it use the same font has VSFilter.

I have also removed Uniscribe, which was used to determine if a font has a specific character.
The sole objective of this PR is to enhance the code quality in the Windows FontCollector.

Now, the user can request a font by a fullname or postscript name like said in https://github.com/libass/libass/pull/203. Previously, on windows, the user could only request a family name.

Should I move the method ``normalizeFilePathCase`` to ``Aegisub/libaegisub/windows/fs.cpp``?

Edit:
The commit de1d3e38d19b1fa3a762450bb3c59bc4e13f7e23 ensures that fonts installed via [AddFontResource](https://learn.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-addfontresourcew) are detected by FontCollector. Previously, they were not recognized.

However, please note that this detection only works on Windows 10 or later. It's uncertain whether this behavior is a bug in the Windows API, but the fonts installed via [AddFontResource](https://learn.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-addfontresourcew) aren't added to the IDWriteFontCollection. Consequently, we cannot convert IDWriteFontFace to IDWriteFont which impeding the ability to verify the presence of certain characters.

moi15moi committed:
[src\meson.build] Add DirectWrite has dependency

moi15moi committed:
[src\font_file_lister_gdi] Rework GDI FontCollector to use DirectWrite

This replaces all the logic of using the Windows registry to obtain the font path by using DirectWrite. The goal is simply to improve the quality of the code. This doesn't change any functionality

moi15moi committed:
[src\meson.build] Remove Uniscribe has dependency

Uniscribe was only used for the FontCollector. Since we now use DirectWrite, we don't need it anymore.

moi15moi committed:
[src\dialog_fonts_collector] Catch exceptions that FontCollector may raise

On Windows, the initialization of the FontCollector can raise an exception

moi15moi committed:
[src\font_file_lister] Document the exception that GdiFontFileLister can throw

moi15moi committed:
[src\font_file_lister_gdi] Correct possible memory leak when an error occur

moi15moi commented:
The CI failure for MacOS seems unrelated to this PR.
```
/usr/bin/hdiutil detach /dev/disk2 -force
hdiutil: couldn't eject "disk2" - Resource busy
FAILED: meson-internal__osx-build-dmg 
env MESON_SOURCE_ROOT=/Users/runner/work/Aegisub/Aegisub MESON_BUILD_ROOT=/Users/runner/work/Aegisub/Aegisub/build MESON_SUBDIR=packages 'MESONINTROSPECT=/Users/runner/hostedtoolcache/Python/3.12.1/x64/bin/meson introspect' /Users/runner/work/Aegisub/Aegisub/tools/osx-dmg.sh /Users/runner/work/Aegisub/Aegisub /Users/runner/work/Aegisub/Aegisub/build 3.2.2
```

moi15moi committed:
Fix error caused by AddFontResource on Windows 10 or higher

arch1t3cht commented on src/font_file_lister_gdi.cpp:
```diff
@@ -210,90 +175,112 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
+
+	// DirectWrite always return the filepath in upper case. Ex: "C:\WINDOWS\FONTS\ARIAL.TTF"
+	DWORD normalized_path_length;
+	if (!normalizeFilePathCase(&path, &normalized_path_length)) {
```
It'd be more consistent to use `path_sh` rather than `&path` here (and similarly in the `ag::fs::path` call below).

moi15moi replied:
Addressed in commit 38a5a99af420aedf926cb6b08cc54c24c8e6d2ad
Now, I don't use WCHAR anymore, so it make this remark obsolete.

arch1t3cht reviewed (changes requested):
I finally had time to look through this - thanks a lot for the PR!

Looking through it, everything seems to be correct as long as I trust you on the DirectWrite/GDI wrangling (and barring the `IDWriteLocalFontLoader` question on Discord). I do have a few comments regarding general code style and conventions, though.

moi15moi committed:
[meson.build] Replace add_project_arguments with conf.set for HAVE_DWRITE_3

moi15moi committed:
[src\dialog_fonts_collector] Update message error and optimisation

moi15moi committed:
[src\font_file_lister_gdi] Correct documentation typo

moi15moi committed:
[src\font_file_lister_gdi] Cosmetic nit - Initialize hfont in one line

moi15moi committed:
[src\font_file_lister_gdi] Cosmetic nit - Remove if statements brace

moi15moi committed:
[src\font_file_lister_gdi] Replace WCHAR param of normalizeFilePathCase to std::wstring

moi15moi commented on meson.build:
```diff
@@ -327,6 +327,10 @@ if host_machine.system() == 'windows'
 endif
 
+if host_machine.system() == 'windows' and cc.has_header('dwrite_3.h')
+    add_project_arguments('-DHAVE_DWRITE_3', language: 'cpp')
```
Addressed in commit 7d099da0994ec37d3584e6c9ce4242f6555f7e8e

moi15moi commented on src/dialog_fonts_collector.cpp:
```diff
@@ -94,7 +94,16 @@ void FontsCollectorThread(AssFile *subs, agi::fs::path const& destination, FcMod
+		catch (agi::EnvironmentError const& err) {
+			AppendText(fmt_tl("* An error occurred when collecting the font: %s.\n", err.GetMessage()), 2);
+			collector->AddPendingEvent(wxThreadEvent(EVT_COLLECTION_DONE));
+			return;
```
Addressed in commit 969758825d452fe347202b91b8b633c1e4925235

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,168 +16,130 @@
+///          - Returns true if the normalization is successful.
+///          - Returns false if an error occurs during file handling or path normalization.
+bool normalizeFilePathCase(WCHAR** path, DWORD* length) {
+	if (path == nullptr || length == nullptr) {
```
Addressed in commit 38a5a99af420aedf926cb6b08cc54c24c8e6d2ad
Now, I don't use WCHAR anymore, so it make this remark obsolete.

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,168 +16,130 @@
+	// We need to convert it to ``fully qualified DOS Path``. Ex: "C:\Windows\Fonts\ariali.ttf"
+	// There isn't any public API that remove the prefix (there is RtlNtPathNameToDosPathName, but it is really hacky to use it)
+	// See: https://stackoverflow.com/questions/31439011/getfinalpathnamebyhandle-result-without-prepended
+	// Even cython remove the prefix manually: https://github.com/python/cpython/blob/963904335e579bfe39101adf3fd6a0cf705975ff/Lib/ntpath.py#L733-L793
```
Addressed in commit 5212d1c7ef686de30f863de0f72cb57208f851c0

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,168 +16,130 @@
+/// @brief Normalize the case of a file path.
+///        Ex: For "C:\WINDOWS\FONTS\ARIAL.TTF", it would return "C:\Windows\Fonts\arial.ttf"
+/// @param[out] path   The path to be normalized. It can be a directory or a file.
+/// @param[out] length Length of the normalized file path string, not including the terminated NULL character.
```
> Or, even better, the input and output could be replaced with some higher-level C++ class like `std::wstring`.

Addressed in commit 38a5a99af420aedf926cb6b08cc54c24c8e6d2ad
Now, I use ``std::wstring``.

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,168 +16,130 @@
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l2992
 	LOGFONTW lf{};
-	lf.lfCharSet = DEFAULT_CHARSET;
+	lf.lfCharSet = DEFAULT_CHARSET; // we should use the one specified in the ass file
```
I wanted to talk about that. In some very specific situation, if we don't use the charset requested by the user, the selected font can be different then the font displayed by VSFilter.
But, this can hardly be emulated on macOS and Linux. So, do we really care to use the charset specified by the user? Even libass doesn't properly emulate how the charset works: https://github.com/libass/libass/issues/662

arch1t3cht replied:
Yes, it's probably better to treat encodings different from 1 as "unsupported". What I really meant here, though, is that currently this comment is a bit confusing. You can remove it or reword it to something like "note that this currently ignores the font encoding specified in the ass file".

moi15moi replied:
Addressed in commit 17a3f487f2479070be45fb454c44dc3dd8d724cc

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,168 +16,130 @@
 CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, int bold, bool italic, std::vector<int> const& characters) {
 	CollectionResult ret;
 
+	int weight = bold == 0 ? 400 :
```
Simply for clarity.

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -210,90 +175,112 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
+	if (hfont == nullptr) {
+		return ret;
 	}
+	agi::scoped_holder<HFONT> hfont_sh(hfont, [](HFONT p) { DeleteObject(p); });
```
Addressed in commit 9af7c22052d59c884022c4875304ee6d9cfd722e

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -210,90 +175,112 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
+	agi::scoped_holder<IDWriteFontFace*> font_face_sh(font_face, [](IDWriteFontFace* p) { p->Release(); });
+
+#ifdef HAVE_DWRITE_3
+	// Fonts added via the AddFontResource API are not included in the IDWriteFontCollection.
```
No, the old font collector wouldn't find them since the fonts added by AddFontResource aren't registered in the Windows Registry.

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,168 +16,130 @@
+{
+	IDWriteFactory* dwrite_factory;
+	if (FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), reinterpret_cast<IUnknown**>(&dwrite_factory))))
+	{
```
Addressed in commit 863d42723d2790aa96330b899e1e20da63c89bf8

moi15moi commented:
> Looking through it, everything seems to be correct as long as I trust you on the DirectWrite/GDI wrangling (and barring the `IDWriteLocalFontLoader` question on Discord).

I did a bunch of test with the user.
You can see my conclusion [here](https://github.com/moi15moi/FindSystemFontsFilename/commit/890301cb79f3c27df5d0a0d04a77ef9349db3900).

moi15moi commented on src/font_file_lister_gdi.cpp:
```diff
@@ -210,90 +149,85 @@ CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, in
+	if (FAILED(local_loader_sh->GetFilePathLengthFromKey(font_file_reference_key, font_file_reference_key_size, &path_length)))
+		return ret;
+
+	WCHAR* path = new WCHAR[path_length + 1];
```
I forgot to release this variable.
I will need to use agi::scoped_holder.

moi15moi replied:
Addressed in commit 62813a6650fe5d645aaf29048d07bd9586662839

moi15moi committed:
[src\font_file_lister_gdi] Replace WCHAR by std::wstring

moi15moi commented:
In some specific cases, the FontCollector can make a mistake in obtaining the font weight and italic style.
For example, install "AliviaRegular_Weight31961.ttf" and "Alivia.otf," which you can find [here](https://github.com/libass/libass/issues/437#issuecomment-1925944380) in the .zip file.

FontCollector will return the font path to "AliviaRegular_Weight31961.ttf" (which is correct), but it will select the weight and italic style from "Alivia.otf." Because of this problem, the ``fake_bold`` and ``fake_italic`` doesn't always represent the true information.

I tried to see if there is a way to avoid this problem, but I didn't find one.
Here is what I tried:

1. Always using the first logfont returned by EnumFontFamiliesEx. I thought the first logfont returned by the Enum is the one that would be selected in the HDC, but it isn't the case. A simple test with Arial shows that it won't necessarily use the first logfont.
2. GetTextMetricsW and GetOutlineTextMetricsW. They return the "simulated" font. For example, if you ask for a font with weight 700, but there is only a 400 font installed, both methods will return metrics with weight 700.
3. IDWriteFont::GetWeight and IDWriteFont::GetStyle. In some specific cases, DirectWrite is incompatible with GDI, and it won't return the same italic value. The same problem occurs with the weight. See this issue: https://github.com/moi15moi/FontCollector/issues/8#issuecomment-1755885113

Conclusion:
There isn't any function that GDI offers to know the italic and weight of the selected font in the HDC.
The only "perfect" solution would be to parse the font like GDI. We could use IDWriteFontFace::TryGetFontTable and get the OS/2 table. If a font doesn't contain an OS/2 table, we would need to use the head table. I don't know if we really want to do that.

moi15moi committed:
[src\font_file_lister_gdi] Use IDWriteFontFace::GetSimulations to detect fake_italic/fake_bold

See this comment: https://github.com/arch1t3cht/Aegisub/pull/107#issuecomment-1975229652

moi15moi committed:
[src\font_file_lister_gdi] If Win7/8 has Win 10 SDK on compile time, correctly verify if font has character(s)

With the Visual Studio 2019 toolchain on Windows 7, it installs the Windows 10 SDK by default. Because of this, ``HAVE_DWRITE_3`` is true, so the ``QueryInterface`` always fails. Now, if the ``QueryInterface`` fails, we try to verify if the font has characters with a Windows Vista SP2 compatible code.

moi15moi committed:
[src\font_file_lister_gdi] Support facename that contains only whitespace AND truncated facename

Problem 1:
Previously, if a user wrote "\fn   ", it would return the font Arial, which is not what we want. This is because when we request EnumFontFamiliesEx with whitespace or an empty lfFaceName, it will enumerate all the installed fonts.

Solution 1:
To resolve this issue, let's implement a solution similar to libass to determine if the selected facename exists: https://github.com/libass/libass/blob/649a7c2e1fc6f4188ea1a89968560715800b883d/libass/ass_directwrite.c#L737-L747

Problem 2:
GDI truncates font names to 31 characters. See: https://github.com/libass/libass/issues/459
However, since I changed the method to determine if a facename exists, I ensured that we still support this "feature".

To test this, I used the font in: https://github.com/libass/libass/issues/710

arch1t3cht commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,284 +16,195 @@
-	return files;
-}
+	std::wstring selected_name(LF_FACESIZE - 1, L'\0');
+	if (!GetTextFaceW(dc_sh, LF_FACESIZE, &selected_name[0]))
```
Instead of `&selected_name[0]` it's better to use `selected_name.data()` here. Apart from just being more idiomatic this is technically required for correctness since `GetTextFaceW` will implicitly override the null terminator of `selected_name`, which the standard explicitly allows for `selected_name.data() + selected_name.size()` and `selected_name[selected_name.size()]` but technically not for `&selected_name[0] + selected_name.size()` (at least if I read the standard correctly. See also https://stackoverflow.com/a/41310431).

moi15moi replied:
I tried it, but I get this error:
```
../src/font_file_lister_gdi.cpp(120): error C2664: 'int GetTextFaceW(HDC,int,LPWSTR)' : impossible de convertir l'argument 3 de 'const _Elem *' en 'LPWSTR'
        with
        [
            _Elem=wchar_t
        ]
```

I'm pretty sure it is because we use [``cpp_std=c++14``](https://github.com/arch1t3cht/Aegisub/blob/cd2e894634bd6fe4038861cfc10491e4ea320123/meson.build#L4). It is only since c++17 or higher that we can use ``selected_name.data()`` in this situation. See https://stackoverflow.com/a/12742517/15835974
I guess that ``&selected_name[0]`` may depend on a undefined behaviour in this situation?

arch1t3cht replied:
Ah, yes. The `data` overload returning a non-const pointer was only added in C++17. I missed that.

However, this is a different question from what standard it'd be legal to use `data` this way in. For that, I was originally looking at cppreference, which listed C++11 for [operator[]](https://en.cppreference.com/w/cpp/string/basic_string/operator_at) and no standard requirement at all (so effectively C++17) for [data](https://en.cppreference.com/w/cpp/string/basic_string/data). But looking at the actual (final drafts of the) standards, it appears that cppreference is wrong and the clause for `operator[]` was indeed added with C++17 and the clause for `data` was added for C++20.

So as far as I can see there's no way to make this technically correct in C++11 without the "pad the string by a byte and resize it afterwards hack". However, seeing as TypesettingTools/Aegisub already switched to C++20 and since this is very likely to still work fine in practice even on earlier standards, I'd be fine with keeping this as it is and adding a comment like `FIXME: This will override the string's terminator, which is not technically correct. After switching to C++20 this should use .data().`

moi15moi replied:
Addressed in commit 44b81fdfae59a9096c6dcab94cf38d6d38e1052d

arch1t3cht commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,284 +16,195 @@
-	// No fonts actually matched
-	if (ret.paths.empty())
+	std::wstring path(path_length, L'\0');
+	if (FAILED(local_loader_sh->GetFilePathFromKey(font_file_reference_key, font_file_reference_key_size, &path[0], path_length + 1)))
```
Again it's better to use `path.data()` here.

arch1t3cht commented on src/font_file_lister_gdi.cpp:
```diff
@@ -16,284 +16,195 @@
-			cb(to_wx(e.GetMessage() + "\n"), 3);
+	// If the selected_name is different then the lf.lfFaceName,
+	// it means that the requested font doesn't exist.
+	if (_wcsnicmp(&selected_name[0], lf.lfFaceName, LF_FACESIZE))
```
And then this could also be `selected_name.data()` for consistency (but this time it's just cosmetic).

arch1t3cht reviewed (commented):
Thanks for the updates. Just a couple more nitpicks left. Once those are addressed, you can squash everything onto one commit and I'll merge it (or if you want to keep the commit history on your branch I'll squash it myself when merging).

moi15moi committed:
[src\font_file_lister_gdi] Add a FIXME comment regarding the utilization of std::wstring over WCHAR

moi15moi committed:
[src\font_file_lister_gdi] Add FIXME comment about charset

moi15moi commented:
> Thanks for the updates. Just a couple more nitpicks left. Once those are addressed, you can squash everything onto one commit and I'll merge it (or if you want to keep the commit history on your branch I'll squash it myself when merging).

I will let you squash them. I would like to conserve the commit. Some of them contain relevant information that may be useful if a problem is discovered in the future.

arch1t3cht commented:
Thanks! Now squashed and merged in [7543060](https://github.com/arch1t3cht/Aegisub/commit/7543060f1d8f008b3220850da7ad51dcb0cc11b0).
